### ISSUE TITLE
[WASI] Should we overlook or restrict the access modes for stdin/stdout/stderr?

### ISSUE THREAD

**Issue by lum1n0us** [bytecodealliance/wasm-micro-runtime#3686]
We've observed that in certain situations, the access modes for STDIN, STDOUT and STDERR do not behave as anticipated. For example, STDOUT may be read-only, while STDIN may be write-only.

It's important that we look into why these unexpected access modes are occurring. However, I'm wondering if we need to take steps to block these incorrect settings in *core/iwasm/libraries/libc-wasi*

Specifically, when determining file descriptor rights in the function `fd_determine_type_rights()`, should we assign fixed and unchangeable rights to STDIN, STDOUT and STDERR?

**Comment by wenyongh** [comment]
I saw the PR #3694 sets the STDIN rights with:
```
(__WASI_RIGHT_FD_ADVISE | __WASI_RIGHT_FD_FILESTAT_GET | \
  __WASI_RIGHT_FD_READ | __WASI_RIGHT_POLL_FD_READWRITE)
```
not sure will it disable the write right for STDIN and disable the read right for STDOUT? It seems the similar issues were reported before, but after discussion with @yamt, we decided not to change the code:
https://github.com/bytecodealliance/wasm-micro-runtime/issues/2645#issuecomment-1790387978
https://github.com/bytecodealliance/wasm-micro-runtime/pull/2606

**Comment by yamt** [comment]
> We've observed that in certain situations, the access modes for STDIN, STDOUT and STDERR do not behave as anticipated. For example, STDOUT may be read-only, while STDIN may be write-only.
> 
> It's important that we look into why these unexpected access modes are occurring. However, I'm wondering if we need to take steps to block these incorrect settings in _core/iwasm/libraries/libc-wasi_

yes, please investigate the situation further before making changes to libc-wasi.

**Comment by lum1n0us** [comment]
The problem originated from integrating WAMR with runwasi, which aims to run a WebAssembly runtime within containerd. I haven't pinpointed the exact issue yet, but there are some unanticipated behaviors occurring within their sandbox environment. Like, `fcntl(STDOUT, F_GETFL, 0)` returns `O_RDONLY`. The unusual result will [strip off write bits](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L518-L519). 

Then I made an trial on linux.
``` bash
$ cat demo.c
int main() {
    int ret;

    ret = fcntl(STDOUT_FILENO, F_GETFL);
    printf("fcntl(stdout, F_GETFL)=%d\n", ret);

    ret = fcntl(STDOUT_FILENO, F_SETFL, O_RDONLY);
    perror("fcntl(stdout, F_SETFL, O_RDONLY)");

    ret = fcntl(STDOUT_FILENO, F_GETFL);
    printf("fcntl(stdout, F_GETFL)=%d\n", ret);

    return EXIT_SUCCESS;
}
$ clang -o demo demo.c
$ demo
fcntl(stdout, F_GETFL)=2
fcntl(stdout, F_SETFL, O_RDONLY): Success
fcntl(stdout, F_GETFL)=2
```

IIUC, this suggests that the access mode for STDOUT cannot be altered. As a result, it appears that the access mode for STDOUT ought to remain fixed, making any previous removal of write permissions unnecessary.

If everyone is in consensus that the access modes for STDIN, STDOUT, and STDERR are fixed and cannot be altered, then we can begin to discuss the suitable combinations of permissions for these three streams.

**Comment by wenyongh** [comment]
Agree that the access modes for STDIN, STDOUT and STDERR should be fixed, and suggest adding `__WASI_RIGHT_FD_READ | __WASI_RIGHT_FD_WRITE` to the rights of them all.

**Comment by yamt** [comment]
> Then I made an trial on linux.

this experiment doesn't seem make sense.
i don't think you can alter O_RDONLY with F_SETFL on any descriptors.

**Comment by lum1n0us** [comment]
> i don't think you can alter O_RDONLY with F_SETFL on any descriptors.

The trial is about asserting that the access mode of STDXXX should be the default value set by the OS and should remain unchanged. `fd_determine_type_rights()` should not use the result of `os_file_get_access_mode()` to strip either read or write bits. Additionally, considering STDXXX has a known file type, there is no need to involve `os_fstat()` and the subsequent `switch-case`. We can simply provide fixed rights for STDXXX. But seems I forgot `__wasi_filetype_t *type` is an output parameter

**Comment by yamt** [comment]
> > i don't think you can alter O_RDONLY with F_SETFL on any descriptors.
> 
> The trial is about asserting that the access mode of STDXXX should be the default value set by the OS and should remain unchanged. `fd_determine_type_rights()` should not use the result of `os_file_get_access_mode()` to strip either read or write bits.

why not?

> Additionally, considering STDXXX has a known file type, there is no need to involve `os_fstat()` and the subsequent `switch-case`. We can simply provide fixed rights for STDXXX. But seems I forgot `__wasi_filetype_t *type` is an output parameter

it's simpler to avoid special cases for fd 0-2 if possible.

**Comment by lum1n0us** [comment]
> why not?

Like you mentioned before, *can't alter O_RDONLY with F_SETFL*. Therefore there will be an assumption that access mode returned by `os_file_get_access_mode()` is able to be trusted(because they are configured by OS), `fd_determine_type_rights()` should use that to strip either read or write bits(like OS will allow write into STDOUT and read from STDIN). But the unexpected result in our case suggest other possibilities that OS might not set those access modes correctly(like OS will not allow write into STDOUT and read from STDIN).

**Comment by lum1n0us** [comment]
> it's simpler to avoid special cases for fd 0-2 if possible.

I think that is what we did in PR.

**Comment by yamt** [comment]
> > why not?
> 
> Like you mentioned before, _can't alter O_RDONLY with F_SETFL_. Therefore there will be an assumption that access mode returned by `os_file_get_access_mode()` is able to be trusted(because they are configured by OS), `fd_determine_type_rights()` should use that to strip either read or write bits(like OS will allow write into STDOUT and read from STDIN). But the unexpected result in our case suggest other possibilities that OS might not set those access modes correctly(like OS will not allow write into STDOUT and read from STDIN).

i guess you need to investigate "the unexpected result in our case" further to pinpoint the cause.
maybe it's a bug of the OS.  (do you mean the embedder, not the OS?)
or maybe it's the intended behavior.
anyway, if the embedder gives us an unreadable descriptor, the best thing we can do is to report the fact to wasm module.

**Comment by yamt** [comment]
> > it's simpler to avoid special cases for fd 0-2 if possible.
> 
> I think that is what we did in PR.

which PR?

https://github.com/bytecodealliance/wasm-micro-runtime/pull/3694 introduced special cases, didn't it?

**Comment by lum1n0us** [comment]
> introduced special cases, didn't it?

Yes, #3694 introduced fixed access modes for STDXXX by ignoring the result of `os_file_get_access_mode()` and `st_filetype`.

> anyway, if the embedder gives us an unreadable descriptor, the best thing we can do is to report the fact to wasm module.

The point is whether you agree that STDIN and STDOUT access mode should not be altered. If so, the process of stripping read or write bits based on the access mode returned by `os_file_get_access_mode()` is not required for STDXXX. Then we should *avoid special cases for file descriptors 0-2 if possible*.

**Comment by yamt** [comment]
> > introduced special cases, didn't it?
> 
> Yes, [#3694](https://github.com/bytecodealliance/wasm-micro-runtime/pull/3694) introduced fixed access modes for STDXXX by ignoring the result of `os_file_get_access_mode()` and `st_filetype`.
> 
> > anyway, if the embedder gives us an unreadable descriptor, the best thing we can do is to report the fact to wasm module.
> 
> The point is whether you agree that STDIN and STDOUT access mode should not be altered. If so, the process of stripping read or write bits based on the access mode returned by `os_file_get_access_mode()` is not required for STDXXX. Then we should _avoid special cases for file descriptors 0-2 if possible_.

i guess i don't agree.
for example, if the embedder gives a write-only host descriptor for wasi stdin, there is no point to set read rights bit for the corresponding wasi descriptor.

i wonder from where you got the idea fd 0-2 should have fixed access modes.
it depends, even on posix environment.
eg. when you run a command on a terminal, fd 0-2 are normally with O_RDWR.
eg. "./a.out 0>foo" gives you stdin with O_WRONLY.

**Comment by lum1n0us** [comment]
```
$ cat demo.c
int main() {
    int ret;

    ret = fcntl(STDOUT_FILENO, F_GETFL);
    printf("fcntl(stdout, F_GETFL)=%d\n", ret);

    ret = fcntl(STDOUT_FILENO, F_SETFL, O_RDONLY);
    perror("fcntl(stdout, F_SETFL, O_RDONLY)");

    ret = fcntl(STDOUT_FILENO, F_GETFL);
    printf("fcntl(stdout, F_GETFL)=%d\n", ret);

    return EXIT_SUCCESS;
}

$ clang -o demo demo.c
$ demo
fcntl(stdout, F_GETFL)=2 // <- O_RDWR
fcntl(stdout, F_SETFL, O_RDONLY): Success  // <- set access mode to other, and success
fcntl(stdout, F_GETFL)=2  // <- STILL BE O_RDWR. So, not-touchable?
```

**Comment by yamt** [comment]
> ```
> $ cat demo.c
> int main() {
>     int ret;
> 
>     ret = fcntl(STDOUT_FILENO, F_GETFL);
>     printf("fcntl(stdout, F_GETFL)=%d\n", ret);
> 
>     ret = fcntl(STDOUT_FILENO, F_SETFL, O_RDONLY);
>     perror("fcntl(stdout, F_SETFL, O_RDONLY)");
> 
>     ret = fcntl(STDOUT_FILENO, F_GETFL);
>     printf("fcntl(stdout, F_GETFL)=%d\n", ret);
> 
>     return EXIT_SUCCESS;
> }
> 
> $ clang -o demo demo.c
> $ demo
> fcntl(stdout, F_GETFL)=2 // <- O_RDWR
> fcntl(stdout, F_SETFL, O_RDONLY): Success  // <- set access mode to other, and success
> fcntl(stdout, F_GETFL)=2  // <- STILL BE O_RDWR. So, not-touchable?
> ```

i agree you can't modify O_RDONY etc with F_SETFL.
it doesn't mean eg. "stdout is always writable" though.

**Comment by yamt** [comment]
let me try to summarize the situation.

* O_RDONLY etc can't/shouldn't be modified with F_SETFL on posix or wasi. i think we all agree on this.
* on posix platforms, sometimes stdio file descriptors have a bit wired access. eg. stdin can be write-only. (see https://github.com/bytecodealliance/wasm-micro-runtime/issues/3686#issuecomment-3010747490 for examples.)
* @yamt believes wasi should just follow posix in this regard and thus proposed https://github.com/bytecodealliance/wasm-micro-runtime/pull/4448.
* @lum1n0us expressed some concerns on it. but @yamt has never understood it.

**Comment by lum1n0us** [comment]
How about this: bypass the part of "strip off read/write bits based on the access mode"?

``` patch
diff --git a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
index 3d90811b..1475ccc6 100644
--- a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
+++ b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c
@@ -465,24 +465,6 @@ fd_determine_type_rights(os_file_handle fd, __wasi_filetype_t *type,
     struct __wasi_filestat_t buf;
     __wasi_errno_t error;
 
-    if (os_is_stdin_handle(fd)) {
-        *rights_base = RIGHTS_STDIN;
-        *rights_inheriting = RIGHTS_STDIN;
-        return __WASI_ESUCCESS;
-    }
-
-    if (os_is_stdout_handle(fd)) {
-        *rights_base = RIGHTS_STDOUT;
-        *rights_inheriting = RIGHTS_STDOUT;
-        return __WASI_ESUCCESS;
-    }
-
-    if (os_is_stderr_handle(fd)) {
-        *rights_base = RIGHTS_STDERR;
-        *rights_inheriting = RIGHTS_STDERR;
-        return __WASI_ESUCCESS;
-    }
-
     error = os_fstat(fd, &buf);
     if (error != __WASI_ESUCCESS)
         return error;
@@ -530,6 +512,11 @@ fd_determine_type_rights(os_file_handle fd, __wasi_filetype_t *type,
             return __WASI_EINVAL;
     }
 
+    if (os_is_stdin_handle(fd) || os_is_stdout_handle(fd)
+        || os_is_stderr_handle(fd)) {
+        return __WASI_ESUCCESS;
+    }
+
     wasi_libc_file_access_mode access_mode;
     error = os_file_get_access_mode(fd, &access_mode);
 
~
```
